src/runtime-rs/crates/hypervisor/src/device/pci_path.rs

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) 2019-2023 Alibaba Cloud
+// Copyright (c) 2019-2023 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use anyhow::{anyhow, Context, Result};
+use lazy_static::lazy_static;
+use regex::Regex;
+
+// get pci path from bdf, and the bdf will be start with "0000:00"
+lazy_static! {
+    static ref RE: Regex = Regex::new(r"^0000:00:([[:xdigit:]]{2}).[[:xdigit:]]$").unwrap();
+}
+
+// Tips:
+// The Re-write `PciSlot` and `PciPath` with rust that it origins from `pcipath.go`:
+//
+
+// The PCI spec reserves 5 bits for slot number (a.k.a. device
+// number), giving slots 0..31
+const PCI_SLOT_BITS: u32 = 5;
+const MAX_PCI_SLOTS: u32 = (1 << PCI_SLOT_BITS) - 1;
+
+// A PciSlot describes where a PCI device sits on a single bus
+//
+// This encapsulates the PCI slot number a.k.a device number, which is
+// limited to a 5 bit value [0x00..0x1f] by the PCI specification
+//
+// To support multifunction device's, It's needed to extend
+// this to include the PCI 3-bit function number as well.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct PciSlot(pub u8);
+
+impl PciSlot {
+    pub fn convert_from_string(s: &str) -> Result<PciSlot> {
+        if s.is_empty() || s.len() > 2 {
+            return Err(anyhow!("string given is invalid."));
+        }
+
+        let base = 16;
+        let n = u64::from_str_radix(s, base).context("convert string to number failed")?;
+        if n >> PCI_SLOT_BITS > 0 {
+            return Err(anyhow!(
+                "number {:?} exceeds MAX:{:?}, failed.",
+                n,
+                MAX_PCI_SLOTS
+            ));
+        }
+
+        Ok(PciSlot(n as u8))
+    }
+
+    pub fn convert_from_u32(v: u32) -> Result<PciSlot> {
+        if v > MAX_PCI_SLOTS {
+            return Err(anyhow!("value {:?} exceeds MAX: {:?}", v, MAX_PCI_SLOTS));
+        }
+
+        Ok(PciSlot(v as u8))
+    }
+
+    pub fn convert_to_string(&self) -> String {
+        format!("{:02x}", self.0)
+    }
+}
+
+// A PciPath describes where a PCI sits in a PCI hierarchy.
+//
+// Consists of a list of PCI slots, giving the slot of each bridge
+// that must be traversed from the PCI root to reach the device,
+// followed by the slot of the device itself.
+//
+// When formatted into a string is written as "xx/.../yy/zz". Here,
+// zz is the slot of the device on its PCI bridge, yy is the slot of
+// the bridge on its parent bridge and so forth until xx is the slot
+// of the "most upstream" bridge on the root bus.
+//
+// If a device is directly connected to the root bus, which used in
+// lightweight hypervisors, such as dragonball/firecracker/clh, and
+// its PciPath.slots will contains only one PciSlot.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct PciPath {
+    // list of PCI slots
+    pub slots: Vec<PciSlot>,
+}
+
+impl PciPath {
```

Comment:
Again, could you implement the standard traits to make this more "rusty" ;)

Reply:
Thx @jodh-intel for your suggestion! Yeah! make it more rusty!

Reply:
> Rustifying the code is definitely a good idea and I think it could have been put in separate patches on top of the original one which was only doing code movement. For both easier review (less changes per patch) and for educational purpose (highlight the improvement in git log).
>
Ok, Thx. I'll fix it.

Reply:
Thx @gkurz Fixed, the following commit to refactor and re-implement the pci path CC @jodh-intel